### ISSUE TITLE
Hypothesis in a section does not obey implicit binders

### ISSUE THREAD

**Issue by jashug** [coq/coq#9363]
#### Version
coq master 137ff25a30fa56bce9160d7ea53a552bb1712dc5

#### Description of the problem

Inside a section, `Hypothesis` and `Variable` do not obey implicit binders, while `Axiom` (and `Parameter`) do. Changing this seems to improve consistency and be potentially useful.

```
(* Outside a section, Hypothesis, Variable, Axiom all obey implicit binders *)
Hypothesis foo1 : forall {n : nat}, True.
Variable foo1' : forall {n : nat}, True.
Axiom foo1'' : forall {n : nat}, True.
Check foo1 (n := 1).
Check foo1' (n := 1).
Check foo1'' (n := 1).

Section bar.
(* Inside a section, Hypothesis and Variable do not, but they should *)
Hypothesis foo2 : forall {n : nat}, True.
Variable foo2' : forall {n : nat}, True.
Axiom foo2'' : forall {n : nat}, True.
Fail Check foo2 (n := 1).
Fail Check foo2' (n := 1).
Check foo2'' (n := 1).
End bar.
```

**Comment by andres-erbsen** [comment]
`Context` is also affected:

```coq
Context (H : forall {n:nat}, n=n -> True).
About H.
(* H : forall n : nat, n = n -> True *)
```

Does #9364 fix that as well?

Fixing that will likely require changes in fiat-crypto and such, but I believe `Arguments` can be used to make these in a backwards-compatible way.

**Comment by jashug** [comment]
No, `Context` is not addressed by #9364, but probably should be addressed.
I'll open a separate issue.
